creme/linear_model/adpredictor.py

```diff
@@ -0,0 +1,175 @@
+
+    """
+    def __init__(self, beta, n_features, prior_probability=None):
+        self.beta = beta
+        self.n_features = n_features
+        self.prior_probability = prior_probability
+        self.mean = 0 if prior_probability is None else stats.norm.ppf(prior_probability) * (beta ** 2 + n_features)
+        self.variance = 1.0
+
+    def __str__(self):
+        return f'Unbiased normal prior on weights 𝒩(μ={self.mean:.3f}, σ2={self.variance:.3f})' if self.prior_probability is None else f'Biased normal prior on weights 𝒩(μ={self.mean:.3f}, σ2={self.variance:.3f})'
+
+    def __call__(self):
+        return self
+
+
+class BOPR:
+    """Bayesian Online Probit Regression Model.
+
+    Parameters:
+        initializer (linear_model.adpredictor.NormalPrior): The initializer used to set beliefs over the weights.
+        surprise (float): Clipping value of the learning step function parameter ensure numerical stability.
+        epsilon (float): Noise parameter of the dynamic correction process.
+
+    Attributes:
+        weights (collections.defaultdict): The current gaussian belief over the weights.
+        
+    """
+    def __init__(self, initializer, surprise, epsilon):
+        self.surprise = surprise
+        self.epsilon = epsilon
+        self.beta = initializer.beta
+        self.prior_probability = initializer.prior_probability
+        self.initializer = initializer
+        self.weights = collections.defaultdict(initializer)
+
+    def _active_mean_variance(self, x):
+        means = [self.weights[i].mean for i, xi in x.items()]
+        variances = [self.weights[i].variance for i, xi in x.items()]
+
+        return sum(means), sum(variances) + self.beta ** 2
+
+    def _gaussian_correction(self, t):
+        """TODO 
+        """
+        t = utils.math.clamp(t, minimum=-self.surprise, maximum=self.surprise)
+        v = stats.norm.pdf(t)/stats.norm.cdf(t)
+        w = v * (v + t)
+        return (v, w)
+
+    def _apply_dynamics(self, weight):
+        """TODO
+        """
+        prior = self.initializer
+        adjusted_variance = weight.variance * prior.variance / ((1.0 - self.epsilon) * prior.variance + self.epsilon * weight.variance)
+        adjusted_mean = adjusted_variance * ((1.0 - self.epsilon) * weight.mean / weight.variance + self.epsilon * prior.mean / prior.variance)
+        prior.variance = adjusted_variance
+        prior.mean = adjusted_mean
+        return prior
+
+    @staticmethod
+    def _target_encoding(y):
+        assert isinstance(y, numbers.Number)
+        return 1.0 if y == 1 else -1.0
+
+
+class AdPredictor(BOPR, base.BinaryClassifier):
+    """AdPredictor is a bayesian online probit regression model devised for search advertising click through rate (CTR) prediction.
+
+    AdPredictor
+
+    Parameters:
+        initializer (linear_model.adpredictor.NormalPrior): The initializer used to set beliefs over the weights.
```

Comment:
With the new documentation system, types are specified as type annotations in the `__init__` method, rather than in the docstring.

Reply:
Ok I will switch to that